behaviour/graql/language/define.feature

```diff
@@ -1018,92 +928,695 @@ Feature: Graql Define Query
+      };
+      """
+
+
+  Scenario: define rule with an undefined attribute set in `then` throws on commit
+    Given graql define throws
+      """
+      define
+      boudicca-is-1960-years-old sub rule,
+      when {
+        $person isa person, has name "Boudicca";
+      }, then {
+        $person has age 1960;
+      };
+      """
+
+
+  Scenario: define rule with an attribute set in `then` on a type that can't have that attribute throws on commit
+    Given graql define throws
+      """
+      define
+      age sub attribute, value long;
+      boudicca-is-1960-years-old sub rule,
+      when {
+        $person isa person, has name "Boudicca";
+      }, then {
+        $person has age 1960;
+      };
+      """
+
+
+  @ignore
+  # TODO: re-enable when rules with attribute values set in `then` that don't match their type throw on commit
+  Scenario: define rule with an attribute value set in `then` that doesn't match the attribute's type throws on commit
+    Given graql define
+      """
+      define
+      nickname sub name;
+      person has nickname;
+      """
+    Given the integrity is validated
+    Then graql define throws
+      """
+      define
+      may-has-nickname-5 sub rule,
+      when {
+        $p has name "May";
+      }, then {
+        $p has nickname 5;
+      };
+      """
+
+
+  Scenario: define rule that infers a relation whose type is undefined throws on commit
+    Then graql define throws
+      """
+      define
+      bonnie-and-clyde-are-partners-in-crime sub rule,
+      when {
+        $bonnie isa person, has name "Bonnie";
+        $clyde isa person, has name "Clyde";
+      }, then {
+        (criminal: $bonnie, sidekick: $clyde) isa partners-in-crime;
+      };
+      """
+
+
+  Scenario: define rule that infers a relation with an incorrect roleplayer throws on commit
+    Then graql define throws
+      """
+      define
+      partners-in-crime sub relation, relates criminal, relates sidekick;
+      person plays criminal;
+      bonnie-and-clyde-are-partners-in-crime sub rule,
+      when {
+        $bonnie isa person, has name "Bonnie";
+        $clyde isa person, has name "Clyde";
+      }, then {
+        (criminal: $bonnie, sidekick: $clyde) isa partners-in-crime;
+      };
+      """
+
+  @ignore
+  # TODO: re-enable when rules cannot infer abstract relations
+  Scenario: define rule that infers an abstract relation throws on commit
+    Then graql define throws
+      """
+      define
+      partners-in-crime sub relation, abstract, relates criminal, relates sidekick;
+      person plays criminal, plays sidekick;
+      bonnie-and-clyde-are-partners-in-crime sub rule,
+      when {
+        $bonnie isa person, has name "Bonnie";
+        $clyde isa person, has name "Clyde";
+      }, then {
+        (criminal: $bonnie, sidekick: $clyde) isa partners-in-crime;
+      };
+      """
+
+  Scenario: define rule that infers an abstract attribute throws on commit (?)
```

Comment:
Right, so this one (along with a number of mutate tests) is unimplemented because I wanted to get on to migrating the grakn integration tests. And the (?) is because I don't currently know what the behaviour will be. My guess is it will, in fact, not throw on commit and turn out to be a bug.

But to be honest, it's not a lot of effort to write, might as well do it

Reply:
Done
```
@ignore
  # TODO: re-enable when rules cannot infer abstract attribute values
  Scenario: define rule that infers an abstract attribute value throws on commit
    Then graql define throws
      """
      define
      number-of-devices sub attribute, value long, abstract;
      person has number-of-devices;
      karl-is-allergic-to-technology sub rule,
      when {
        $karl isa person, has name "Karl";
      }, then {
        $karl has number-of-devices 0;
      };
      """
```